### PR TITLE
build: Require libc++-16 or later

### PR DESCRIPTION
**Pull request by maflcko**
I am not familiar with macOS, but given that https://github.com/bitcoin/bitcoin/pull/28622 prescribes a minimum of libc++-16 for macOS, doing the same on Linux could make sense?

This is needed for stuff:

* `std::ranges::equal`, see https://github.com/bitcoin/bitcoin/pull/29071
* `std::source_location`, see https://github.com/bitcoin/bitcoin/pull/28999#issuecomment-1851949647

I don't think anyone is using `-stdlib=libc++` on Linux, except for the CI and OSS-Fuzz?

The minimum required clang version remains at `clang++-15`.

### CODE REVIEW

File: doc/dependencies.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ You can find installation instructions in the `build-*.md` file for your platfor
 | --- | --- |
 | [Autoconf](https://www.gnu.org/software/autoconf/) | [2.69](https://github.com/bitcoin/bitcoin/pull/17769) |
 | [Automake](https://www.gnu.org/software/automake/) | [1.13](https://github.com/bitcoin/bitcoin/pull/18290) |
-| [Clang](https://clang.llvm.org) | [13.0](https://github.com/bitcoin/bitcoin/pull/28210) |
+| [Clang](https://clang.llvm.org) | [13.0](https://github.com/bitcoin/bitcoin/pull/28210) (with at least libstdc++-10, or at least [libc++ 16.0](https://github.com/bitcoin/bitcoin/pull/29077))|
```
Comment (fanquake) [inline doc/dependencies.md:11]: Given https://github.com/bitcoin/bitcoin/pull/28687#issuecomment-1855569550 and https://godbolt.org/z/n4oTE9sTE the `libstdc++-10` requirement isn't going to work here? Looks like this really needs to be clang-16.

**Comment by fanquake** [comment]
> and OSS-Fuzz?

So won't this break as soon as we have code that requires libstdc++-16, given that OSS-Fuzz is still pinned to 15?

Are there no other projects using (some?) C++20 code on OSS-Fuzz?

File: doc/dependencies.md (hunk 1 above)
Comment (theuni) [reply doc/dependencies.md:11]: More missing in v10: https://github.com/bitcoin/bitcoin/pull/29081#issuecomment-1856419623

**Comment by fanquake** [comment]
> So I'd say to close this and bump to 16 wholesale?

I think generally, that would be more straightforward, then supporting varying compiler + std lib combinations, and seems possible to do at this point.

**Review by SamyDgz — APPROVED** [review]

**Pull request closed without merging**